### PR TITLE
SAA-1887: Allocate to a session running today

### PR DESCRIPTION
**Pull request by munslowl**
Allow a new allocation to happen for a session(s) today indicated by the user choosing the scheduled instance in the UI.
Similar for update an allocation.
If prisoner can attend the sessions then new activities will be created and messages posted for synch.
This can also work for sessions earlier in the day should the user dawdle in the UI

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt [hunk 1]
```diff
@@ -242,7 +242,7 @@ data class ActivitySchedule(
     prisonerNumber: PrisonerNumber,
     payBand: PrisonPayBand?,
     bookingId: Long,
-    startDate: LocalDate = LocalDate.now(),
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt:245]: When you allocate you can now allocate from today

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt [hunk 2]
```diff
@@ -352,12 +352,12 @@ data class Allocation(
 
   fun isEnded() = status(PrisonerStatus.ENDED)
 
-  fun updateExclusion(slot: ActivityScheduleSlot, daysOfWeek: Set<DayOfWeek>): Exclusion? {
+  fun updateExclusion(slot: ActivityScheduleSlot, daysOfWeek: Set<DayOfWeek>, startDate: LocalDate): Exclusion? {
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt:355]: Can't assume when the start date is any more.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityScheduleService.kt [hunk 3]
```diff
@@ -144,7 +145,11 @@ class ActivityScheduleService(
   fun allocatePrisoner(scheduleId: Long, request: PrisonerAllocationRequest, allocatedBy: String) {
     log.info("Allocating prisoner ${request.prisonerNumber}.")
 
-    require(request.startDate!! > LocalDate.now()) { "Allocation start date must be in the future" }
+    val today = LocalDate.now()
+
+    require(request.startDate!! >= today) { "Allocation start date must not be in the past" }
+
+    if (request.startDate == today && request.nextAvailableScheduleInstanceId == null) throw IllegalArgumentException("The next session must be provided when allocation start date is today")
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityScheduleService.kt:152]: If the user chooses to allocate from today then we need the id of the next session as we may have just missed it's start time.

[Confirmed this is okay](https://mojdt.slack.com/archives/C04E1PDMTNH/p1720689081453259?thread_ts=1720687708.126279&cid=C04E1PDMTNH)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt [hunk 4]
```diff
@@ -96,6 +97,37 @@ class ManageAttendancesService(
...
+
+    val nextAvailableInstance = scheduledInstanceRepository.findOrThrowNotFound(nextAvailableScheduleInstanceId)
+
+    require(nextAvailableInstance.activitySchedule == allocation.activitySchedule) {
+      "Allocation does not belong to same activity schedule as selected instance"
+    }
+
+    // Need to create attendances for today?
+    return allocation.activitySchedule.instances().filter {
+      it.sessionDate == LocalDate.now() &&
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ManageAttendancesService.kt:123]: Create attendance if:

- It's today
- Its a session on or later than the selected session
- They are allowed to attend

File: src/test/resources/test_data/seed-activity-id-1.sql [hunk 5]
```diff
@@ -46,17 +46,27 @@ values (4, 2, 'A11111A', 10001, 3, '2022-10-12', null, '2022-10-10 10:00:00', 'M
...
 
+insert into allocation(allocation_id, activity_schedule_id, prisoner_number, booking_id, prison_pay_band_id, start_date, end_date, allocated_time, allocated_by, deallocated_time, deallocated_by, deallocated_reason, suspended_time, suspended_by, suspended_reason, prisoner_status)
+values (6, 2, 'A1234BC', 10004, 3, current_timestamp + interval '2 day', null, '2022-10-10 10:00:00', 'MRS BLOGS', null, null, null, null, null, null, 'ACTIVE');
+
 insert into scheduled_instance(activity_schedule_id, session_date, start_time, end_time, cancelled, cancelled_time, cancelled_by, cancelled_reason, comment)
 values (1, '2022-10-10', '10:00:00', '11:00:00', false, null, null, null, null);
 
-insert into attendance(attendance_id, scheduled_instance_id, prisoner_number, attendance_reason_id, comment, recorded_time, recorded_by, status, pay_amount, bonus_amount, pieces)
-values (1, 1, 'A11111A', null, null, null, null, 'WAITING', null, null, null);
+insert into scheduled_instance(activity_schedule_id, session_date, start_time, end_time, cancelled, cancelled_time, cancelled_by, cancelled_reason, comment)
```
Comment (munslowl) [inline src/test/resources/test_data/seed-activity-id-1.sql:55]: If you insert with including pk then you can screw the nextval

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityScheduleService.kt [hunk 6]
```diff
@@ -144,7 +145,11 @@ class ActivityScheduleService(
   fun allocatePrisoner(scheduleId: Long, request: PrisonerAllocationRequest, allocatedBy: String) {
     log.info("Allocating prisoner ${request.prisonerNumber}.")
 
-    require(request.startDate!! > LocalDate.now()) { "Allocation start date must be in the future" }
+    val today = LocalDate.now()
+
+    require(request.startDate!! >= today) { "Allocation start date must not be in the past" }
+
+    if (request.startDate == today && request.scheduleInstanceId == null) throw IllegalArgumentException("The next session must be provided when allocation start date is today")
```
Comment (munslowl) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityScheduleService.kt:152]: If the user chooses to allocate from today then we need the id of the next session as we may have just missed it's start time.

[Confirmed this is okay](https://mojdt.slack.com/archives/C04E1PDMTNH/p1720689081453259?thread_ts=1720687708.126279&cid=C04E1PDMTNH)

**Review by timoj58 — APPROVED** [review]

**Pull request merged**